src/cargo/core/source/mod.rs

```diff
@@ -44,6 +44,9 @@ pub trait Source {
     /// Ensure that the source is fully up-to-date for the current session on the next query.
     fn invalidate_cache(&mut self);
 
+    /// If quiet, the source should not display any progress or status messages.
+    fn set_quiet(&mut self, quiet: bool);
```

Comment:
Wait, don't we cache sources generally (and the "wait" logic has to workaround that)?  I feel like mutating a cached item like this is pretty brittle (in an already brittle system) and I'd hope we can find a way to avoid making it worse.

Reply:
I'm not sure I follow. Sources are inherently mutable. I'm not sure I would say they are "cached" so much as they are lazily created as needed and held in the `SourceMap`. But by their nature they are accessed from the map and mutated as needed.

Reply:
As for the caching / lazy loading, I don't remember the details but I do remember running into a lot of problems implementing this because I was getting stale state due to caches that said "everything is loaded, no reason to talk to the server again".

>  But by their nature they are accessed from the map and mutated as needed.

From my quick glance, they are mutated as-needed in that you invalidate their cache and they reload.  I didn't see any other real "state" to them that could leak from one operation to another.  That is what I'm concerned about.  As I said, I ran into a lot of problems with cargo assuming everything was only done once because commands are transient but when a command needed to do things multiple times, they broke.  While I am not refreshed on all the details to map out whether its actually safe or not, this feels like one of those things that could easily cause people problems in the future.